anvil/plot.py

```diff
@@ -0,0 +1,193 @@
+def plot_2pf(training_geometry, two_point_function):
+    corr = np.empty((training_geometry.length, training_geometry.length))
+    std = np.empty((training_geometry.length, training_geometry.length))
+    for t in range(training_geometry.length):
+        for x in range(training_geometry.length):
+            corr[x, t] = float(two_point_function(t, x)[0])
+            std[x, t] = float(bootstrap(two_point_function(t, x)))
+
+    fractional_std = std / corr
+
+    fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(13, 6), sharey=True)
+    ax2.set_title(r"$\sigma_G / G$")
+    ax1.set_title("$G(x, t)$")
+    ax1.set_xlabel("$x$")
+    ax2.set_xlabel("$x$")
+    ax1.set_ylabel("$t$")
+    im1 = ax1.pcolor(corr)
+    im2 = ax2.pcolor(fractional_std)
+
+    ax1.yaxis.set_major_locator(MaxNLocator(integer=True))
+    ax1.xaxis.set_major_locator(MaxNLocator(integer=True))
+    ax2.xaxis.set_major_locator(MaxNLocator(integer=True))
+
+    fig.colorbar(im1, ax=ax1)
+    fig.colorbar(im2, ax=ax2)
+
+    return fig
+
+
+@figure
+def plot_volume_averaged_2pf(volume_averaged_2pf):
+    fig, ax = plt.subplots()
+    ax.set_title("Volume-averaged two point function")
+    ax.set_ylabel("$G_k(0,0)$")
+    ax.set_xlabel("$t$")
+    ax.plot(volume_averaged_2pf(0, 0), "-")
+    return fig
+
+
+@figure
+def plot_autocorrelation_2pf(autocorrelation_2pf):
+    """Plot autocorrelation as a function of Monte Carlo time,  and
+    integrated autocorrelation, exponential autocorrelation, and the "g" function
+    whose minimum corresponds to a window size where errors in the integrated
+    autocorrelation are minimised."""
+    autocorrelation, tau_int_W, tau_exp_W, g_W, W_opt = autocorrelation_2pf
+    W = np.arange(1, tau_int_W.size + 1)
+
+    fig, (ax1, ax2, ax3, ax4) = plt.subplots(4, figsize=(6, 10))
+    # Autocorrelation
+    ax1.set_title("Autocorrelation of volume-averaged two point function")
+    ax1.set_ylabel("$\Gamma_{G(s)}(t)$")
+    ax1.set_xlabel("$t$")
+    ax1.plot(autocorrelation)
+    ax1.plot([W_opt + 1,] * 2, [autocorrelation.min(), autocorrelation.max()], "r-")
+    # Integrated autocorrelation time
+    ax2.set_title("Integrated autocorrelation time")
+    ax2.set_ylabel(r"$\tau_{int}(W)$")
+    ax2.plot(W, tau_int_W)
+    ax2.plot([W_opt,] * 2, [tau_int_W.min(), tau_int_W.max()], "r-")
+    # Exponential autocorrelation time
+    ax3.set_title("Exponential autocorrelation time")
+    ax3.set_ylabel(r"$\tau_{exp}(W)$")
+    ax3.plot(W, tau_exp_W)
+    ax3.plot([W_opt,] * 2, [tau_exp_W.min(), tau_exp_W.max()], "r-")
+    # "g" function
+    ax4.set_title("g")
+    ax4.set_ylabel("$g$")
+    ax4.set_xlabel("$W$")
+    ax4.plot(W, g_W)
+    ax4.plot([W_opt,] * 2, [g_W.min(), g_W.max()], "r-", label=r"$W_{opt}$")
+    ax4.legend()
+
+    return fig
+
+
+def plot_bootstrap_dist(observable, label):
+    """Plot the distribution of some observable calculated using many bootstrap samples"""
+
+    def do_plot(ax, full_data, bootstrap_data, std):
```

Comment:
does this need to be inside the function? I didn't check super thoroughly but I don't think it's using any of the functionality of a nested function?